File: src/routes/contacts.ts
import { and, desc, eq, inArray, isNull, lt, or, sql } from 'drizzle-orm'
import { Hono } from 'hono'
import { z } from 'zod'
import { makeDb } from '../db'
import { type Contact, contacts } from '../db/schema'
import type { Env } from '../env'
import { recordEvent } from '../lib/audit'
import { HTTPError } from '../lib/errors'
import { PaginationQuery, buildListResponse, decodeCursor } from '../lib/pagination'
import { type AppVars, requireAuth } from '../middleware/auth'

const ContactCreate = z.object({
  external_id: z.string().max(255).optional(),
  first_name: z.string().max(255).optional(),
  last_name: z.string().max(255).optional(),
  email: z.string().email().max(320).optional(),
  phone: z.string().max(64).optional(),
  title: z.string().max(255).optional(),
  company_id: z.string().uuid().optional(),
  tags: z.array(z.string()).optional(),
  data: z.record(z.unknown()).optional(),
})

const ContactUpdate = ContactCreate.partial()

const ContactListQuery = PaginationQuery.extend({
  q: z.string().optional(),
  email: z.string().optional(),
  company_id: z.string().uuid().optional(),
  include_deleted: z.coerce.boolean().default(false),
})

function serialize(c: Contact) {
  return {
    id: c.id,
    external_id: c.externalId,
    first_name: c.firstName,
    last_name: c.lastName,
    email: c.email,
    phone: c.phone,
    title: c.title,
    company_id: c.companyId,
    tags: c.tags,
    data: c.data,
    created_at: c.createdAt,
    updated_at: c.updatedAt,
    deleted_at: c.deletedAt,
  }
}

export const contactsRouter = new Hono<{ Bindings: Env; Variables: AppVars }>()
contactsRouter.use('*', requireAuth)

// --- List ---
contactsRouter.get('/', async (c) => {
  const q = ContactListQuery.parse(Object.fromEntries(new URL(c.req.url).searchParams))
  const db = makeDb(c.env.DB)
  const wsId = c.var.principal.workspace.id

  const whereClauses = [eq(contacts.workspaceId, wsId)]
  if (!q.include_deleted) whereClauses.push(isNull(contacts.deletedAt))
  if (q.email) whereClauses.push(eq(contacts.email, q.email))
  if (q.company_id) whereClauses.push(eq(contacts.companyId, q.company_id))

  if (q.q) {
    // FTS for fuzzy search across first_name / last_name / email / title.
    const fts = await c.env.DB.prepare(
      'SELECT contact_id FROM contacts_fts WHERE workspace_id = ? AND contacts_fts MATCH ? LIMIT 500',
    )
      .bind(wsId, q.q)
      .all<{ contact_id: string }>()
    const ids = fts.results.map((r: { contact_id: string }) => r.contact_id)
    if (ids.length === 0) return c.json({ items: [], next_cursor: null })
    whereClauses.push(inArray(contacts.id, ids))
  }

  if (q.cursor) {
    const pos = decodeCursor(q.cursor)
    if (pos) {
      whereClauses.push(
        or(
          lt(contacts.createdAt, new Date(pos.createdAt)),
          and(eq(contacts.createdAt, new Date(pos.createdAt)), lt(contacts.id, pos.id)),
        )!,
      )
    }
  }

  const rows = await db
    .select()
    .from(contacts)
    .where(and(...whereClauses))
    .orderBy(desc(contacts.createdAt), desc(contacts.id))
    .limit(q.limit + 1)
    .all()

  return c.json(buildListResponse(rows, q.limit, serialize))
})

// --- Create ---
contactsRouter.post('/', async (c) => {
  const body = ContactCreate.parse(await c.req.json())
  const db = makeDb(c.env.DB)
  const wsId = c.var.principal.workspace.id

  if (body.external_id) {
    const existing = await db
      .select()
      .from(contacts)
      .where(and(eq(contacts.workspaceId, wsId), eq(contacts.externalId, body.external_id)))
      .get()
    if (existing) throw new HTTPError(409, 'external_id already used in this workspace')
  }

  const created = await db
    .insert(contacts)
    .values({
      workspaceId: wsId,
      externalId: body.external_id ?? null,
      firstName: body.first_name ?? null,
      lastName: body.last_name ?? null,
      email: body.email ?? null,
      phone: body.phone ?? null,
      title: body.title ?? null,
      companyId: body.company_id ?? null,
      tags: body.tags ?? [],
      data: body.data ?? {},
    })
    .returning()
    .get()

  await recordEvent(c, {
    eventType: 'contact.created',
    entityType: 'contact',
    entityId: created.id,
  })

  return c.json(serialize(created), 201)
})

// --- Get ---
contactsRouter.get('/:id', async (c) => {
  const db = makeDb(c.env.DB)
  const wsId = c.var.principal.workspace.id
  const row = await db
    .select()
    .from(contacts)
    .where(and(eq(contacts.workspaceId, wsId), eq(contacts.id, c.req.param('id'))))
    .get()
  if (!row || row.deletedAt) throw new HTTPError(404, 'contact not found')
  return c.json(serialize(row))
})

// --- Update ---
contactsRouter.patch('/:id', async (c) => {
  const body = ContactUpdate.parse(await c.req.json())
  const db = makeDb(c.env.DB)
  const wsId = c.var.principal.workspace.id
  const id = c.req.param('id')

  const existing = await db
    .select()
    .from(contacts)
    .where(and(eq(contacts.workspaceId, wsId), eq(contacts.id, id)))
    .get()
  if (!existing || existing.deletedAt) throw new HTTPError(404, 'contact not found')

  const updated = await db
    .update(contacts)
    .set({
      externalId: body.external_id ?? existing.externalId,
      firstName: body.first_name ?? existing.firstName,
      lastName: body.last_name ?? existing.lastName,
      email: body.email ?? existing.email,
      phone: body.phone ?? existing.phone,
      title: body.title ?? existing.title,
      companyId: body.company_id ?? existing.companyId,
      tags: body.tags ?? existing.tags,
      data: body.data ?? existing.data,
    })
    .where(eq(contacts.id, id))
    .returning()
    .get()
  if (!updated) throw new HTTPError(500, 'contact vanished after update')

  await recordEvent(c, {
    eventType: 'contact.updated',
    entityType: 'contact',
    entityId: id,
    payload: { changed: Object.keys(body) },
  })

  return c.json(serialize(updated))
})

// --- Soft delete ---
contactsRouter.delete('/:id', async (c) => {
  const db = makeDb(c.env.DB)
  const wsId = c.var.principal.workspace.id
  const id = c.req.param('id')

  const existing = await db
    .select()
    .from(contacts)
    .where(and(eq(contacts.workspaceId, wsId), eq(contacts.id, id)))
    .get()
  if (!existing || existing.deletedAt) throw new HTTPError(404, 'contact not found')

  await db.update(contacts).set({ deletedAt: new Date() }).where(eq(contacts.id, id)).run()
  await recordEvent(c, {
    eventType: 'contact.deleted',
    entityType: 'contact',
    entityId: id,
  })

  return c.body(null, 204)
})

// --- Restore (un-soft-delete) ---
contactsRouter.post('/:id/restore', async (c) => {
  const db = makeDb(c.env.DB)
  const wsId = c.var.principal.workspace.id
  const id = c.req.param('id')

  const existing = await db
    .select()
    .from(contacts)
    .where(and(eq(contacts.workspaceId, wsId), eq(contacts.id, id)))
    .get()
  if (!existing) throw new HTTPError(404, 'contact not found')
  if (!existing.deletedAt) return c.json(serialize(existing))

  const restored = await db
    .update(contacts)
    .set({ deletedAt: null })
    .where(eq(contacts.id, id))
    .returning()
    .get()
  await recordEvent(c, {
    eventType: 'contact.restored',
    entityType: 'contact',
    entityId: id,
  })
  return c.json(serialize(restored!))
})

// Drizzle's `sql` import is referenced indirectly via `inArray`; keep
// import marker so the bundler keeps it.
void sql

File: legacy/app/routers/contacts.py
from __future__ import annotations

from datetime import UTC, datetime

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query
from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session

from app.db import get_db
from app.deps import Pagination, Principal, get_pagination, get_principal, require_role
from app.models import Contact, EntityType, MemberRole
from app.schemas import (
    BulkUpsertResult,
    ContactIn,
    ContactMergeRequest,
    ContactMergeResponse,
    ContactOut,
    ContactPatch,
    Page,
)
from app.services.diffs import compute_changes
from app.services.duplicates import find_duplicates
from app.services.duplicates import serialize as serialize_duplicates
from app.services.events import emit
from app.services.merge import merge_contacts
from app.services.pagination import apply_cursor, encode_cursor

router = APIRouter(prefix="/contacts", tags=["contacts"])


def _base_query(db: Session, workspace_id: str, include_deleted: bool):
    q = select(Contact).where(Contact.workspace_id == workspace_id)
    if not include_deleted:
        q = q.where(Contact.deleted_at.is_(None))
    return q


@router.get("", response_model=Page[ContactOut])
def list_contacts(
    db: Session = Depends(get_db),
    p: Principal = Depends(get_principal),
    page: Pagination = Depends(get_pagination),
    q: str | None = Query(None, description="substring match on first/last name/email"),
    email: str | None = None,
    company_id: str | None = None,
    tag: str | None = None,
    include_deleted: bool = False,
):
    query = _base_query(db, p.workspace.id, include_deleted)
    if q:
        like = f"%{q.lower()}%"
        query = query.where(
            or_(
                func.lower(Contact.first_name).like(like),
                func.lower(Contact.last_name).like(like),
                func.lower(Contact.email).like(like),
            )
        )
    if email:
        query = query.where(func.lower(Contact.email) == email.lower())
    if company_id:
        query = query.where(Contact.company_id == company_id)
    if tag:
        query = query.where(Contact.tags.contains([tag]))

    total = db.scalar(select(func.count()).select_from(query.subquery())) or 0
    query = apply_cursor(query, model=Contact, cursor=page.cursor)
    query = query.order_by(Contact.created_at.desc(), Contact.id.desc()).limit(page.limit + 1)

    rows = db.scalars(query).all()
    next_cursor = None
    if len(rows) > page.limit:
        last = rows[page.limit - 1]
        next_cursor = encode_cursor(last.created_at, last.id)
        rows = rows[: page.limit]
    return Page[ContactOut](
        items=[ContactOut.model_validate(r) for r in rows],
        next_cursor=next_cursor,
        count=total,
    )


@router.post("", response_model=ContactOut, status_code=201)
def create_contact(
    payload: ContactIn,
    background: BackgroundTasks,
    db: Session = Depends(get_db),
    p: Principal = Depends(get_principal),
) -> ContactOut:
    c = Contact(workspace_id=p.workspace.id, **payload.model_dump())
    db.add(c)
    try:
        db.flush()
    except Exception as e:  # noqa: BLE001
        db.rollback()
        raise HTTPException(status_code=409, detail=f"conflict: {e.__class__.__name__}")
    emit(
        db,
        p,
        event_type="contact.created",
        entity_type=EntityType.contact,
        entity_id=c.id,
        payload={"contact_id": c.id},
        background=background,
    )
    db.commit()
    db.refresh(c)
    return ContactOut.model_validate(c)


# Static sub-paths must come before the ``{contact_id}`` route — FastAPI
# matches in declaration order and would otherwise treat "duplicates" or
# "merge" as a contact id and 500 on UUID parse.
@router.get("/duplicates")
def list_duplicates(
    min_score: float = Query(0.7, ge=0.0, le=1.0),
    limit: int = Query(100, ge=1, le=500),
    name_threshold: float = Query(0.80, ge=0.0, le=1.0),
    db: Session = Depends(get_db),
    p: Principal = Depends(get_principal),
) -> dict:
    """Return likely duplicate contact pairs in the current workspace.

    Each pair carries a ``score`` (0–1) and a ``reason`` explaining why they
    matched:

    - ``exact_email`` (1.0): same email, case-insensitive
    - ``name_similar_same_company`` (0.8): full-name trigram similarity above
      ``name_threshold`` AND both linked to the same company
    - ``last_name_same_first_variant`` (0.7): identical last name AND first
      names either match exactly or one is a prefix of the other (Tom/Thomas)

    Feed each pair into ``POST /contacts/merge`` when you've confirmed they're
    the same person.
    """
    pairs = find_duplicates(
        db,
        p.workspace.id,
        min_score=min_score,
        limit=limit,
        name_threshold=name_threshold,
    )
    return {"items": serialize_duplicates(pairs), "count": len(pairs)}


@router.get("/{contact_id}", response_model=ContactOut)
def get_contact(contact_id: str, db: Session = Depends(get_db), p: Principal = Depends(get_principal)):
    c = db.get(Contact, contact_id)
    if not c or c.workspace_id != p.workspace.id:
        raise HTTPException(status_code=404, detail="not found")
    return ContactOut.model_validate(c)


@router.patch("/{contact_id}", response_model=ContactOut)
def patch_contact(
    contact_id: str,
    payload: ContactPatch,
    background: BackgroundTasks,
    db: Session = Depends(get_db),
    p: Principal = Depends(get_principal),
) -> ContactOut:
    c = db.get(Contact, contact_id)
    if not c or c.workspace_id != p.workspace.id:
        raise HTTPException(status_code=404, detail="not found")
    updates = payload.model_dump(exclude_unset=True)
    for k, v in updates.items():
        setattr(c, k, v)
    changes = compute_changes(c, list(updates.keys()))
    emit(
        db,
        p,
        event_type="contact.updated",
        entity_type=EntityType.contact,
        entity_id=c.id,
        payload={"changes": changes},
        background=background,
    )
    db.commit()
    db.refresh(c)
    return ContactOut.model_validate(c)


@router.delete("/{contact_id}")
def delete_contact(
    contact_id: str,
    background: BackgroundTasks,
    db: Session = Depends(get_db),
    p: Principal = Depends(get_principal),
    hard: bool = False,
):
    c = db.get(Contact, contact_id)
    if not c or c.workspace_id != p.workspace.id:
        raise HTTPException(status_code=404, detail="not found")
    if hard:
        db.delete(c)
    else:
        c.deleted_at = datetime.now(UTC)
    emit(
        db,
        p,
        event_type="contact.deleted",
        entity_type=EntityType.contact,
        entity_id=c.id,
        payload={"hard": hard},
        background=background,
    )
    db.commit()
    return {"ok": True}


@router.post("/bulk_upsert", response_model=BulkUpsertResult)
def bulk_upsert(
    items: list[ContactIn],
    background: BackgroundTasks,
    db: Session = Depends(get_db),
    p: Principal = Depends(get_principal),
) -> BulkUpsertResult:
    created = updated = 0
    ids: list[str] = []
    for item in items:
        existing = None
        if item.external_id:
            existing = db.scalar(
                select(Contact).where(
                    Contact.workspace_id == p.workspace.id,
                    Contact.external_id == item.external_id,
                )
            )
        elif item.email:
            existing = db.scalar(
                select(Contact).where(
                    Contact.workspace_id == p.workspace.id,
                    func.lower(Contact.email) == item.email.lower(),
                )
            )
        if existing:
            for k, v in item.model_dump(exclude_unset=True).items():
                setattr(existing, k, v)
            updated += 1
            ids.append(existing.id)
            emit(
                db,
                p,
                event_type="contact.updated",
                entity_type=EntityType.contact,
                entity_id=existing.id,
                payload={"via": "bulk_upsert"},
                background=background,
            )
        else:
            c = Contact(workspace_id=p.workspace.id, **item.model_dump())
            db.add(c)
            db.flush()
            created += 1
            ids.append(c.id)
            emit(
                db,
                p,
                event_type="contact.created",
                entity_type=EntityType.contact,
                entity_id=c.id,
                payload={"via": "bulk_upsert"},
                background=background,
            )
    db.commit()
    return BulkUpsertResult(created=created, updated=updated, ids=ids)


@router.post("/merge", response_model=ContactMergeResponse)
def merge_endpoint(
    req: ContactMergeRequest,
    background: BackgroundTasks,
    db: Session = Depends(get_db),
    p: Principal = Depends(require_role(MemberRole.owner, MemberRole.admin)),
) -> ContactMergeResponse:
    """Merge two duplicate contacts — owner/admin only because it's destructive.

    The ``winner_id`` contact is kept and enriched with non-conflicting fields
    from the ``loser_id``; the loser is soft-deleted with ``data.merged_into``
    pointing at the winner. All references (deal.primary_contact_id,
    relationships, notes/tasks/activities/files/memory_links) are rewritten.

    Pass ``dry_run=true`` to preview. Use ``field_preferences`` to override the
    default "winner wins unless null" rule per field:

        {"field_preferences": {"title": "loser"}}
    """
    try:
        result = merge_contacts(
            db,
            p.workspace.id,
            winner_id=req.winner_id,
            loser_id=req.loser_id,
            field_preferences=req.field_preferences,
            dry_run=req.dry_run,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    if not req.dry_run:
        emit(
            db,
            p,
            event_type="contact.merged",
            entity_type=EntityType.contact,
            entity_id=req.winner_id,
            payload={
                "loser_id": req.loser_id,
                "changes": result.changes,
                "references_rewritten": result.references_rewritten,
            },
            background=background,
        )
        db.commit()
    else:
        db.rollback()
    return ContactMergeResponse(
        winner_id=req.winner_id,
        loser_id=req.loser_id,
        changes=result.changes,
        references_rewritten=result.references_rewritten,
        warnings=result.warnings,
        dry_run=req.dry_run,
    )
